syntax = "proto2";

// Command types
enum CommandTypeEnum{
  LED_test = 0;
  I2C_test = 1;
  SPI_test = 2;
  GPIO_digital = 3;
  Analog_read = 4;		// Analog in, ADC config
  Analog_write = 5;		// Analog out, PWM
}




// I2C options - descriptionben megadni melyik i2c interfeszt konfigolja fel
enum i2cDirection{
	I2C_write = 0;
	I2C_read = 1;
}

enum i2cBus{
	I2C1 = 1;
	I2C2 = 2;
	I2C3 = 3;
}

// Master only
message i2cMessage{
 required i2cBus bus = 1;
 required uint32 address = 2;
 required uint32 register = 3;
 required i2cDirection direction = 4; 	// 1: read, 0: write
}




// SPI options
// only master modes should be available
enum spiMode{
	FULL_DUPLEX_MASTER = 0;		
	FULL_DUPLEX_SLAVE = 1;
	HALF_DUPLEX_MASTER = 2;
	HALF_DUPLEX_SLAVE = 3;
	RECEIVE_ONLY_MASTER = 4;
	RECEIVE_ONLY_SLAVE = 5;
	TRANSMIT_ONLY_MASTER = 6;
	TRANSMIT_ONLY_SLAVE = 7;
}

enum spiDirection{
 SPI_TRANSMIT = 0;
 SPI_RECEIVE = 1;
}

enum clockMode{
	MODE_0 = 0;
	MODE_1 = 1;
	MODE_2 = 2;
	MODE_3 = 3;
}

enum hardwareNSS{
	DISABLE = 0;
	NSS_INPUT = 1;
	NSS_OUTPUT = 2;
}

enum spiBus{
	SPI1 = 1;
	SPI2 = 2;
	SPI3 = 3;
}

message spiMessage{
 required spiBus bus = 1;
 required clockMode mode = 2;	//ide esetleg hintet megjeleniteni
 required uint32 command = 3;
 required spiDirection direction = 4;
 optional uint32 dummyclocks = 5;
}


// digitalGPIO options
enum gpioDirection{
	GPIO_OUTPUT = 0;
	GPIO_INPUT = 1;
}

enum gpioPinState{
	GPIO_LOW = 0;
	GPIO_HIGH = 1;
}

enum gpioPins{
	PA0 = 0;
	PA1 = 1;
	PA2 = 2;
	PA3 = 3;
	PA4 = 4;
	PA5 = 5;	
	PA6 = 6;
	PA7 = 7;
	PA8 = 8;
	PA9 = 9;
	PA10 = 10;
	PA11 = 11;
	PA12 = 12;
	PA13 = 13;	
	PA14 = 14;
	PA15 = 15;	
	PB0 = 20;
	PB1 = 21;
	PB2 = 22;
	PB3 = 23;	
	PB4 = 24;
	PB5 = 25;	
	PB6 = 26;
	PB7 = 27;	
	PB8 = 28;
	PB9 = 29;	
	PB10 = 30;
	PB11 = 31;
	PB12 = 32;
	PB13 = 33;
	PB14 = 34;
	PB15 = 35;
	PC0 = 40;
	PC1 = 41;
	PC2 = 42;
	PC3 = 43;	
	PC4 = 44;
	PC5 = 45;	
	PC6 = 46;
	PC7 = 47;	
	PC8 = 48;
	PC9 = 49;	
	PC10 = 50;
	PC11 = 51;
	PC12 = 52;
	PC13 = 53;
	PC14 = 54;
	PC15 = 55;
	Temp = 60;
	Vrefint = 61;
	Vbat = 62;
}



message digitalGPIO{
 required gpioPins pin = 1;
 required gpioDirection direction = 2; 	 // Input/output
 required gpioPinState state = 3;	// Low, high
}



// Analog IN message
enum adcResolution{		
	ADC_12_BITS = 0;
	ADC_10_BITS = 1;
	ADC_8_BITS = 2;
	ADC_6_BITS = 3;
}


message analogIN{		
 required gpioPins pin = 1;
 required adcResolution resolution = 2;
}



// Analog OUT message
message analogOUT{
 required gpioPins pin = 1;
 required uint32 frequency = 2;
 required uint32 dutyCycle = 3;
}



enum successfulWrite{
 LED_SET = 0;
 I2C_WRITE_OK = 1;
 I2C_WRITE_FAIL = 2;
 SPI_WRITE_OK = 3;
 SPI_WRITE_FAIL = 4;
 GPIO_SET = 5;
 GPIO_RESET = 6;
 GPIO_WRITE_FAIL = 7;
 PWM_SET = 8;
 PWM_RESET = 9;
 PWM_WRITE_FAIL = 10;
}



// Main message
message Command{
 required CommandTypeEnum commandType = 1;
 optional	i2cMessage	i2c = 2;
 optional	spiMessage	spi = 3;
 optional	digitalGPIO	gpio = 4;
 optional analogIN analog_in = 6;
 optional analogOUT	analog_out = 7;
 optional uint32 responseRead = 10;					// Register value
 optional successfulWrite responseWrite = 11;		// Write success
}
